### PR TITLE
Xds client split

### PR DESCRIPTION
**Pull request by larry-safran**
Split off from #11254

### CODE REVIEW

File: census/src/test/java/io/grpc/census/CensusModulesTest.java [hunk 1]
```diff
@@ -136,7 +138,7 @@ public class CensusModulesTest {
       ClientStreamTracer.StreamInfo.newBuilder()
           .setCallOptions(CallOptions.DEFAULT.withOption(NAME_RESOLUTION_DELAYED, 10L)).build();
 
-  private static class StringInputStream extends InputStream {
+  private static class StringInputStream extends InputStream implements KnownLength {
```
Comment (YifeiZhuang) [inline census/src/test/java/io/grpc/census/CensusModulesTest.java:141]: do we need change in this file?

File: xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java [hunk 2]
```diff
@@ -51,6 +51,7 @@ public class SharedXdsClientPoolProviderTest {
   @Rule
   public final ExpectedException thrown = ExpectedException.none();
   private final Node node = Node.newBuilder().setId("SharedXdsClientPoolProviderTest").build();
+  private static final String DUMMY_TARGET = "dummy";
```
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java:54]: We should test at least two targets, and verify, e.g. the second xdsClient is not the same instance as the first one, is ref counted.

File: xds/src/test/java/io/grpc/xds/ControlPlaneRule.java [hunk 3]
```diff
@@ -86,9 +87,15 @@ public class ControlPlaneRule extends TestWatcher {
   private XdsTestControlPlaneService controlPlaneService;
   private XdsTestLoadReportingService loadReportingService;
   private XdsNameResolverProvider nameResolverProvider;
+  private final int port;
 
   public ControlPlaneRule() {
+    this(0);
+  }
+
+  public ControlPlaneRule(int port) {
```
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/ControlPlaneRule.java:96]: This file seems not related to the change. It is used for fallback?

File: xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java [hunk 4]
```diff
@@ -135,12 +135,15 @@ public void run() {
                   new Object[]{value.getResourceNamesList(), value.getErrorDetail()});
               return;
             }
+
             String resourceType = value.getTypeUrl();
-            if (!value.getResponseNonce().isEmpty()
-                && !String.valueOf(xdsNonces.get(resourceType)).equals(value.getResponseNonce())) {
+            if (!value.getResponseNonce().isEmpty() && xdsNonces.containsKey(resourceType)
```
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java:140]: this change also seems not related?

File: xds/src/main/java/io/grpc/xds/CsdsService.java [hunk 5]
```diff
@@ -117,53 +119,77 @@ public void onCompleted() {
...
+    }
+    return ClientStatusResponse.newBuilder().addAllConfig(clientConfigs).build();
+  }
+
   @VisibleForTesting
-  static ClientConfig getClientConfigForXdsClient(XdsClient xdsClient) throws InterruptedException {
+  static ClientConfig getClientConfigForXdsClient(XdsClient xdsClient, String target)
+      throws InterruptedException {
     ClientConfig.Builder builder = ClientConfig.newBuilder()
+        .setClientScope(target)
```
Comment (YifeiZhuang) [inline xds/src/main/java/io/grpc/xds/CsdsService.java:192]: is the clientScope tested?

File: xds/src/test/java/io/grpc/xds/CsdsServiceTest.java [hunk 6]
```diff
@@ -105,7 +106,7 @@ public void setUp() {
       // because true->false return mutation prevents fetchClientStatus from completing the request.
       csdsStub = ClientStatusDiscoveryServiceGrpc
           .newBlockingStub(grpcServerRule.getChannel())
-          .withDeadline(Deadline.after(3, TimeUnit.SECONDS));
+          .withDeadline(Deadline.after(30, TimeUnit.SECONDS));
       csdsAsyncStub = ClientStatusDiscoveryServiceGrpc.newStub(grpcServerRule.getChannel());
     }
```
Comment (YifeiZhuang) [inline xds/src/test/java/io/grpc/xds/CsdsServiceTest.java:112]: Add a test case to verify multiple xdsClients's configs will be dumped by csds would be good.

File: census/src/test/java/io/grpc/census/CensusModulesTest.java (hunk 1 above)
Comment (larry-safran) [reply census/src/test/java/io/grpc/census/CensusModulesTest.java]: You're right, reverted it.

File: xds/src/main/java/io/grpc/xds/CsdsService.java (hunk 5 above)
Comment (larry-safran) [reply xds/src/main/java/io/grpc/xds/CsdsService.java:192]: Added tests

File: xds/src/test/java/io/grpc/xds/ControlPlaneRule.java (hunk 3 above)
Comment (larry-safran) [reply xds/src/test/java/io/grpc/xds/ControlPlaneRule.java]: Yes, reverted it

File: xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java (hunk 4 above)
Comment (larry-safran) [reply xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java:140]: This fix is needed to support anything trying a nonce other than the initial empty value.

File: xds/src/test/java/io/grpc/xds/CsdsServiceTest.java (hunk 6 above)
Comment (larry-safran) [reply xds/src/test/java/io/grpc/xds/CsdsServiceTest.java]: Added a test named multipleXdsClients

File: xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java (hunk 4 above)
Comment (YifeiZhuang) [reply xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java:140]: I still don't see this change related to "support multiple xds client". Looks this file can be reverted without impacting anything?
But I see this potentially avoid NPE when it receives another resource type other than the hardcoded resource type in `xdsNonces` keyset.

File: xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java (hunk 2 above)
Comment (YifeiZhuang) [reply xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java:54]: I see that multiple targets have been covered in csds service test. Still, it would be nice to have multiple targets tested in this UT.

**Review by YifeiZhuang — APPROVED** [review]
LGTM in general with two minor comments.

File: xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java [hunk 7]
```diff
@@ -132,15 +146,19 @@ public XdsClient getObject() {
...
+            xdsClient = new XdsClientImpl(
+                DEFAULT_XDS_TRANSPORT_FACTORY,
+                bootstrapInfo,
+                scheduler,
+                BACKOFF_POLICY_PROVIDER,
+                GrpcUtil.STOPWATCH_SUPPLIER,
+                TimeProvider.SYSTEM_TIME_PROVIDER,
+                MessagePrinter.INSTANCE,
+                new TlsContextManagerImpl(bootstrapInfo));
+          } catch (Exception e) {
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java:159]: What is this catching? I don't see any new exceptions in #11254 either.

**Comment by ejona86** [comment]
You don't need to wait for me to merge this. I wasn't planning on providing a full review. I just poked in and had the drive-by comment.
